hazelcast/src/main/java/com/hazelcast/internal/util/InstantiationUtils.java

```diff
@@ -0,0 +1,138 @@
+                if (selectedConstructor == null) {
+                    selectedConstructor = (Constructor<T>) constructor;
+                } else {
+                    throw new AmbigiousInstantiationException("Class " + clazz + " has multiple constructors matching "
+                            + "given parameters: " + Arrays.toString(params));
+                }
+            }
+        }
+        return selectedConstructor;
+    }
+
+    private static boolean isParamsMatching(Constructor<?> constructor, Object[] params) {
+        Class<?>[] constructorParamTypes = constructor.getParameterTypes();
+        if (constructorParamTypes.length != params.length) {
+            return false;
+        }
+        for (int i = 0; i < constructorParamTypes.length; i++) {
+            Class<?> constructorParamType = constructorParamTypes[i];
+            Object param = params[i];
+            if (constructorParamType.isPrimitive()) {
+                // 1. Constructors can accept primitive arguments
+                // 2. Parameters are never primitives because they are passed inside an object array
+                // 3. As a user I expect this utility class to do auto-unboxing of passed parameters
+                //  ->  If a constructor argument is a primitive type then we have to use its boxed version otherwise
+                //      isAssignableFrom() bellow fails. It's because an instance of Integer.class cannot be directly
+                //      assigned to int.class without Java compiler doing its magic.
+                if (param == null) {
+                    // passed parameter was null, but the argument in constructor is primitive - it's not matching
+                    return false;
+                } else {
+                    constructorParamType = toBoxedType(constructorParamType);
+                }
+            }
+            if (param == null) {
+                // null is matching all non-primitive types
+                continue;
+            }
+            Class<?> paramType = param.getClass();
+            if (!constructorParamType.isAssignableFrom(paramType)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static Class<?> toBoxedType(Class<?> type) {
+        assert type.isPrimitive();
+        if (type == boolean.class) {
+            return Boolean.class;
+        } else if (type == byte.class) {
+            return Byte.class;
+        } else if (type == char.class) {
+            return Character.class;
+        } else if (type == double.class) {
+            return Double.class;
+        } else if (type == float.class) {
+            return Float.class;
+        } else if (type == int.class) {
+            return Integer.class;
+        } else if (type == long.class) {
+            return Long.class;
+        } else if (type == short.class) {
+            return Short.class;
+        } else {
+            // should never happen ;-)
```

Comment:
https://gph.is/2d6uTTr

Reply:
actually I was thinking about throwing `AssertionError`. or `HellBreaksLooseError` here :)

Reply:
actually, I was too quick here. 

```java
Exception in thread "main" java.lang.IllegalArgumentException: Unknown primitive type void
    at JavaClass.toBoxedType(JavaClass.java:44)
    at JavaClass.main(JavaClass.java:20)
```

```java

Exception in thread "main" java.lang.IllegalArgumentException: Unknown primitive type [I
    at JavaClass.toBoxedType(JavaClass.java:44)
    at JavaClass.main(JavaClass.java:20)
```
thanks to @frant-hartm for showing these to me!

I guess I should sign-up for Java 101 on Coursera or something :)